crates/air-parser/src/parser/validator.rs

```diff
@@ -0,0 +1,209 @@
+            FunctionPart::FuncName(func_name) => self.meet_instr_value(func_name, span),
+            FunctionPart::ServiceIdWithFuncName(service_id, func_name) => {
+                self.meet_instr_value(service_id, span);
+                self.meet_instr_value(func_name, span);
+            }
+        }
+    }
+
+    fn meet_args(&mut self, args: &[CallInstrArgValue<'i>], span: Span) {
+        for arg in args {
+            self.meet_instr_arg_value(arg, span);
+        }
+    }
+
+    fn meet_instr_value(&mut self, instr_value: &CallInstrValue<'i>, span: Span) {
+        match instr_value {
+            CallInstrValue::JsonPath { variable, .. } => self.meet_variable(variable, span),
+            CallInstrValue::Variable(variable) => self.meet_variable(variable, span),
+            _ => {}
+        }
+    }
+
+    fn meet_instr_arg_value(&mut self, instr_arg_value: &CallInstrArgValue<'i>, span: Span) {
+        match instr_arg_value {
+            CallInstrArgValue::JsonPath { variable, .. } => self.meet_variable(variable, span),
+            CallInstrArgValue::Variable(variable) => self.meet_variable(variable, span),
+            _ => {}
+        }
+    }
+
+    fn meet_variable(&mut self, name: &'i str, span: Span) {
+        if !contains(&self.met_variables, name, span) && !contains(&self.met_iterable, name, span) {
+            self.unresolved_variables.insert(name, span);
+        }
+    }
+
+    fn meet_call_output(&mut self, call_output: &CallOutputValue<'i>, span: Span) {
+        let variable_name = match call_output {
+            CallOutputValue::Scalar(variable) => variable,
+            CallOutputValue::Accumulator(accumulator) => accumulator,
+            CallOutputValue::None => return,
+        };
+
+        self.met_variables.insert(variable_name, span);
+    }
+
+    fn meet_iterable_value(&mut self, iterable_value: &IterableValue<'i>, span: Span) {
+        match iterable_value {
+            IterableValue::JsonPath { variable, .. } => self.meet_variable(variable, span),
+            IterableValue::Variable(variable) => self.meet_variable(variable, span),
+        }
+    }
+
+    fn meet_iterator(&mut self, iterator: &'i str, span: Span) {
+        self.met_iterable.insert(iterator, span);
+    }
+}
+
+use std::cmp::Ordering;
+impl PartialOrd for Span {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        let self_min = std::cmp::min(self.left, self.right);
+        let other_min = std::cmp::min(other.left, other.right);
+
+        if self_min < other_min {
+            Some(Ordering::Less)
+        } else if self == other {
+            Some(Ordering::Equal)
+        } else {
+            Some(Ordering::Greater)
+        }
+    }
+}
+
+fn contains(multimap: &MultiMap<&str, Span>, key: &str, key_span: Span) -> bool {
+    let found_spans = match multimap.get_vec(key) {
+        Some(found_spans) => found_spans,
+        None => return false,
+    };
+
+    found_spans.iter().any(|s| s < &key_span)
```

Comment:
Maybe only store a minimal found span? I think this would allow to avoid `multimap` using `hashmap` instead, and also would make `contains` a constant time operation.

I mean, on `insert` it is possible to compare new span with existing, and choose the minimum of the two